kubectl proxy: override request host

FabianKramm opened the pull request:
Signed-off-by: fabiankramm <fab.kramm@googlemail.com>

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind bug
<!--
Add one of the following kinds:
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

**What this PR does / why we need it**:

Currently `kubectl proxy` does not work correctly with many API Gateway proxies as backends (e.g. api server behind a nginx ingress controller). The problem behind this is that the `Host` header field that arrives at the backend still points to the `kubectl proxy` host of the original request. 

This is due to the go client preferring the `req.Host` over `req.URL.Host`.

https://github.com/golang/go/blob/a0d6420d8be2ae7164797051ec74fa2a2df466a1/src/net/http/h2_bundle.go#L7947-L7954

However, `httputil.ReverseProxy` and also upgrade path only set `req.URL.Host`. Hence, `req.Host` still points to the proxy host and is preferred by the client over `req.URL.Host` and sent to the backend server as `Host` header. 

https://github.com/golang/go/blob/a0d6420d8be2ae7164797051ec74fa2a2df466a1/src/net/http/httputil/reverseproxy.go#L105-L121

https://github.com/kubernetes/kubernetes/blob/979c644df2ad051015a8ccf6475e3476eda0cea0/staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go#L284-L286

This leads to the request being routed to the localhost again instead of the correct target location in the backend server. This PR fixes this issue by also setting `req.Host` in the request for the upgrade path and non upgrade path.

Relevant go issue: https://github.com/golang/go/issues/28168

Other people experiencing this bug: https://github.com/lensapp/lens/issues/873#issuecomment-692644642

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
none
```

FabianKramm committed:
override request host for kubectl proxy

Signed-off-by: fabiankramm <fab.kramm@googlemail.com>

FabianKramm commented:
/cc @eddiezane

eddiezane commented:
/ok-to-test

liggitt commented:
Thanks for the pull request. Since this is also used server side in many proxying applications, we need to understand all the code paths that reach this location.

To start, can you write a test that demonstrates the bug this is fixing? 

Are there any scenarios in which the currently preferred host is desired?

FabianKramm commented:
@liggitt thanks for taking a look! After some consideration I made this optional in the `UpgradeAwareHandler`, because it might be desired to not override the `Host` http header in other scenarios than `kubectl proxy` (I can't think of a concrete example, but it is probably safer to just use it only for `kubectl proxy` for now). 

After some more testing I also found out this is also an issue for HTTP/1.1 connections, where the wrong `Host` http header arrives at the backend. The problem however only becomes apparent if the backend needs to act on this `Host` header like API gateways or ingress controllers that forward the request somehow themselves. I updated the comments and PR description accordingly. 

To demonstrate what exactly this PR is fixing, I added a new test case to the existing test suite that will fail without the option enabled and succeed with the new option in `UpgradeAwareHandler`. It checks if the correct `Host` http header arrives at the backend.

pacoxu commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -238,6 +240,11 @@ func (h *UpgradeAwareHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
+	if h.OverrideRequestHost {
+		// exchanging req.Host with the backend location is necessary for backends that act on the HTTP host header (e.g. API gateways),
+		// because req.Host has preference over req.URL.Host in filling this header field
+		newReq.Host = h.Location.Host
```
A minor comments here, newReq setters should be inserted after Line 231.
Line 227 - Line 231 for newReq settings.
Line 233 - Line 242 for proxy setings.

> As  req.Host has preference over  req.URL.Host
Why not check whether they are different or newReq.Host is nil before override request host  every time？

FabianKramm replied:
@pacoxu Done!

fedebongio commented:
/triage accepted

dougsland commented:
/cc

FabianKramm commented:
@dougsland any updates on this?

dougsland commented:
> @dougsland any updates on this?

On my queue for tomorrow. Just reviewed your PR 97350.

dougsland commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -282,6 +289,11 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 	} else {
 		klog.V(6).Infof("Connecting to backend proxy (direct dial) %s\n  Headers: %v", &location, clone.Header)
+		if h.OverrideRequestHost {
+			// exchanging req.Host with the backend location is necessary for backends that act on the HTTP host header (e.g. API gateways),
```
I believe no need to duplicate the comment from line 232. git blame also would help identify which issue is about.

FabianKramm replied:
@dougsland done!

dougsland commented on staging/src/k8s.io/kubectl/pkg/proxy/proxy_server.go:
```diff
@@ -195,6 +195,7 @@ func NewServer(filebase string, apiProxyPrefix string, staticPrefix string, filt
 	proxy := proxy.NewUpgradeAwareHandler(target, transport, false, false, responder)
 	proxy.UpgradeTransport = upgradeTransport
 	proxy.UseRequestLocation = true
+	proxy.OverrideRequestHost = true
```
@FabiamKramm It seems a real issue when using api behind nginx as instance (I remember to see a k3s issue related to this). Can we add a flag to enable OverrideRequestHost? I agree that would be safe to not enabled by default (at least for now IMO). However, we will need people like @liggitt and others to approve this one.

dougsland replied:
Let me update my comment:
I am good **not** using flags as kubeproxy goal is to override localhost and so far looks sane to avoid any breakage with current code.

dougsland commented:
> After some more testing I also found out this is also an issue for HTTP/1.1 connections, where the wrong `Host` http 
> header arrives at the backend. The problem however only becomes apparent if the backend needs to act on this
> `Host` header like API gateways or ingress controllers that forward the request themselves. I updated the comments
> and PR description accordingly.
> 

Would be nice if you share the steps here too. To have everyone in the same page.

dougsland commented:
/lgtm

dougsland commented:
Adding more people for review
/cc @rikatz

FabianKramm commented:
> > After some more testing I also found out this is also an issue for HTTP/1.1 connections, where the wrong `Host` http
> > header arrives at the backend. The problem however only becomes apparent if the backend needs to act on this
> > `Host` header like API gateways or ingress controllers that forward the request themselves. I updated the comments
> > and PR description accordingly.
> 
> Would be nice if you share the steps here too. To have everyone in the same page.

Sure, I tested it on docker-desktop for mac, but it should work with every kubernetes cluster where you can retrieve the tls keys:
1. Retrieve the api server tls.key, tls.crt and ca.crt from the VM (you can find them in path `/run/config/pki`)
2. Create the api server tls secret with `kubectl create secret generic api-server-cert --namespace=kube-system --from-file=tls.crt=tls.crt --from-file=tls.key=tls.key --from-file=ca.crt=ca.crt`
3. Create a service to access the api server directly: 
```yaml
apiVersion: v1
kind: Service
metadata:
  name: api-server
  namespace: kube-system
spec:
  selector:
    component: kube-apiserver
  ports:
    - protocol: TCP
      port: 443
      targetPort: 6443
```
4. Install the nginx-ingress controller via kubectl (see https://kubernetes.github.io/ingress-nginx/deploy/#docker-for-mac): 
```
kubectl apply -f https://raw.githubusercontent.com/kubernetes/ingress-nginx/controller-v0.44.0/deploy/static/provider/cloud/deploy.yaml
```
5. Create an ingress for the api server with:
```
apiVersion: networking.k8s.io/v1beta1
kind: Ingress
metadata:
  name: api-server
  namespace: kube-system
  annotations:
    nginx.ingress.kubernetes.io/proxy-ssl-secret: kube-system/api-server-cert
    nginx.ingress.kubernetes.io/backend-protocol: HTTPS
spec:
  tls:
    - hosts:
        - kubernetes.default.svc.cluster.local
      secretName: api-server-cert
  rules:
    - host: kubernetes.default.svc.cluster.local
      http:
        paths:
          - path: /
            backend:
              serviceName: api-server
              servicePort: 443
```
6. Make sure you have added `kubernetes.default.svc.cluster.local` to your `/etc/hosts`, like this:
```
127.0.0.1 kubernetes.default.svc.cluster.local
```
7. Make sure you have a service account token that has enough rights for the demo, if not create a cluster role binding via:
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: cluster-admin-default
subjects:
  - kind: ServiceAccount
    name: default
    namespace: default
roleRef:
  kind: ClusterRole
  name: cluster-admin
  apiGroup: rbac.authorization.k8s.io
```
8. With the service account token, host and the ca.crt you can create a new kube config `config.yaml`, like this:
```yaml
apiVersion: v1
kind: Config
clusters:
  - cluster:
      certificate-authority-data: LS0tLS1...
      server: https://kubernetes.default.svc.cluster.local
    name: nginx-proxy
contexts:
  - context:
      cluster: nginx-proxy
      user: nginx-proxy
    name: nginx-proxy
current-context: nginx-proxy
preferences: {}
users:
  - name: nginx-proxy
    user:
      token: eyJhbG...
```
9. Make sure you can access kubernetes through this kube config via: `kubectl --kubeconfig config.yaml get pods`
10. Now test `kubectl --kubeconfig config.yaml proxy`
11. Navigate to `http://127.0.0.1:8001/` which should give you an 404 error not found because nginx will try to route the request to its `127.0.0.1:8001`
12. Compile kubectl with the changes introduced in this PR, reexecute `kubectl --kubeconfig config.yaml proxy`, navigate to `http://127.0.0.1:8001/` again and you should see all API paths, because nginx know gets the correct host with `kubernetes.default.svc.cluster.local`

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -74,6 +74,8 @@ type UpgradeAwareHandler struct {
 	RequireSameHostRedirects bool
 	// UseRequestLocation will use the incoming request URL when talking to the backend server.
 	UseRequestLocation bool
+	// OverrideRequestHost will override the HTTP host header when talking to the backend server.
+	OverrideRequestHost bool
```
```suggestion
	// UseLocationHost overrides the HTTP host header in requests to the backend server to use the Host from Location
	UseLocationHost bool
```

liggitt replied:
also document how this interacts with the `UseRequestLocation` option (it's not clear to me how it should)

FabianKramm replied:
@liggitt I changed it like you suggested and added the explanation from the [net/http/request.go](https://github.com/golang/go/blob/d4b26382342c98a95b85140b2863bc30c48edd68/src/net/http/request.go#L119-L122) comment about the difference between req.URL.Host (set by `UseRequestLocation`) and req.Host (set by `UseLocationHost`)

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -282,6 +289,9 @@ func (h *UpgradeAwareHandler) tryUpgrade(w http.ResponseWriter, req *http.Reques
 	} else {
 		klog.V(6).Infof("Connecting to backend proxy (direct dial) %s\n  Headers: %v", &location, clone.Header)
+		if h.OverrideRequestHost {
+			clone.Host = h.Location.Host
```
does code coverage show this line being exercised in unit tests?

FabianKramm replied:
@liggitt it was not. I added a new test case called `use location host - invalid upgrade` to `TestServeHTTP`, that covers it and tests this explicitly.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware.go:
```diff
@@ -227,6 +229,11 @@ func (h *UpgradeAwareHandler) ServeHTTP(w http.ResponseWriter, req *http.Request
+	if h.OverrideRequestHost {
+		// exchanging req.Host with the backend location is necessary for backends that act on the HTTP host header (e.g. API gateways),
+		// because req.Host has preference over req.URL.Host in filling this header field
+		newReq.Host = h.Location.Host
```
does code coverage show this line being exercised in unit tests?

FabianKramm replied:
@liggitt that is covered by the new test case `use location host` in `TestServeHTTP `

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/proxy/upgradeaware_test.go:
```diff
@@ -162,6 +164,7 @@ func TestServeHTTP(t *testing.T) {
 		notExpectedRespHeader []string
 		upgradeRequired       bool
 		expectError           func(err error) bool
+		overrideRequestHost   bool
```
```suggestion
		useLocationHost   bool
```

FabianKramm replied:
@liggitt I changed it like you suggested

FabianKramm commented:
@liggitt thanks for the review! I updated the PR with the changes you requested, would be great if you could take a look again

FabianKramm commented:
/retest

FabianKramm commented:
/retest

liggitt commented:
/lgtm
/approve
